pkg/tfbridge/walk.go

```diff
@@ -281,3 +287,394 @@ func LookupSchemaInfoPath(schemaPath SchemaPath, schemaInfo *SchemaInfo) *Schema
+func (r VisitDataSourceRoot) isVisitRoot()             {}
+
+type VisitConfigRoot struct{ token tokens.Type }
+
+func (r VisitConfigRoot) PulumiToken() tokens.Type { return r.token }
+func (r VisitConfigRoot) isVisitRoot()             {}
+
+type Visitor = func(VisitInfo) (VisitResult, error)
+
+// A wrapper around [TraverseSchemas] that panics on error. See [TraverseSchemas] for
+// documentation.
+func MustTraverseSchemas(prov *ProviderInfo, traversalID string, visitor Visitor) {
+	err := TraverseSchemas(prov, traversalID, visitor)
+	contract.AssertNoErrorf(err, "failed to traverse provider schemas")
+}
+
+// _Efficiently_ traverse all schemas in a provider.
+//
+// TraverseSchemas operates in two stages:
+//
+//   - During `make tfgen` time, all fields in the provider are visited. TraverseSchemas
+//     builds a list of fields where `VisitResult{ HasEffect: true }` is returned.
+//
+//   - During the provider's normal runtime, only "effectful" fields are visited.
+//
+// traversalId must be unique for each traversal within a `make tfgen`, but the same
+// between `make tfgen` and the provider runtime. A simple descriptive string like `"apply
+// labels"` works great here.
+func TraverseSchemas(prov *ProviderInfo, traversalID string, visitor Visitor, opts ...TraversalOption) error {
+	prov.MetadataInfo.assertValid() // We must have metadata info for an efficient (sparse) traversal.
+
+	var errs []error
+
+	var traverseOptions traversalOptions
+	for _, opt := range opts {
+		opt(&traverseOptions)
+	}
+
+	// forEffect indicates that the traversal is being run for side effects only, so
+	// we can visit only "effectful" nodes.
+	forEffect := currentRuntimeStage == runningProviderStage
+	if b := traverseOptions.forEffect; b != nil {
+		forEffect = *b
+	}
+
+	var effects traversalRecord
+
+	// We are in forEffect mode, so we need to load in the list of paths where we need
+	// to replay effects.
+	if forEffect {
+		var ok bool
+		var err error
+		effects, ok, err = md.Get[traversalRecord](prov.GetMetadata(), traversalID)
+		contract.Assertf(ok, "could not find traversalID %q in metadata, are you missing a `make tfgen`?", traversalID)
+		if err != nil {
+			return fmt.Errorf("unable to find effect list: %w", err)
+		}
+	}
+
+	rMap := prov.P.ResourcesMap()
+	for tfName, rInfo := range prov.Resources {
+		rShim, ok := rMap.GetOk(tfName)
+		if !ok {
+			continue
+		}
+
+		var err error
+
+		effect := ensureMap(&effects.Resources)[tfName]
+
+		effect, err = traverseResourceOrDataSource(tfName, rShim, rInfo,
+			visitor, forEffect, effect)
+		if err != nil {
+			errs = append(errs, err)
+		}
+		effects.Resources[tfName] = effect
+	}
+
+	dMap := prov.P.DataSourcesMap()
+	for tfName, dInfo := range prov.DataSources {
+		rShim, ok := dMap.GetOk(tfName)
+		if !ok {
+			continue
+		}
+
+		var err error
+		effect := ensureMap(&effects.DataSources)[tfName]
+
+		effect, err = traverseResourceOrDataSource(tfName, rShim, dInfo,
```

Comment:
your variable naming is slightly confusing - you are using `effect` and `effects` both in this function and in `traverseResourceOrDataSource` but in `traverseResourceOrDataSource` you return an `effects` and assign it to `effect`.

It'd be a bit clearer if you have 3 variable names and match up the ones which are the same across function boundaries.

Reply:
The type of `effects` also changes - it is `effects traversalRecord` here but `effects []SchemaPath` in `traverseSchemaMap`

Reply:
perhaps something like `effectRecords` -> `effectPaths` -> `effect`